File: karate-chop/java/src/main/java/com/codekata/karatechop/RecursiveBinarySearch.java
package com.codekata.karatechop;

public class RecursiveBinarySearch implements BinarySearch {

    @Override
    public int chop(int numberToFind, int[] inputNumbers) {
        return binarySearch(inputNumbers, numberToFind, 0, inputNumbers.length - 1);
    }

    private int binarySearch(int[] inputNumbers, int numberToFind, int imin, int imax) {

        if (imax < imin) {
            return NOT_FOUND;
        } else {
            int imid = midpoint(imin, imax);

            if (inputNumbers[imid] > numberToFind) {
                return binarySearch(inputNumbers, numberToFind, imin, imid - 1);
            } else if (inputNumbers[imid] < numberToFind) {
                return binarySearch(inputNumbers, numberToFind, imid + 1, imax);
            } else {
                return imid;
            }
        }
    }

    private int midpoint(int imin, int imax) {
        return (imin + imax) / 2;
    }
}

File: karate-chop/js/components/RecursiveBinarySearch.js
"use strict";

class RecursiveBinarySearch {

    chop(numberToFind, inputNumbers) {

        function binarySearch(inputNumbers, numberToFind, imin, imax) {

            function midpoint(imin, imax) {
                return Number.parseInt((imin + imax) / 2);
            }

            if (imax < imin) {
                return -1;
            } else {

                let imid = midpoint(imin, imax);

                if (inputNumbers[imid] > numberToFind) {
                    return binarySearch(inputNumbers, numberToFind, imin, imid - 1);
                } else if (inputNumbers[imid] < numberToFind) {
                    return binarySearch(inputNumbers, numberToFind, imid + 1, imax);
                } else {
                    return imid;
                }
            }
        }

        return binarySearch(inputNumbers, numberToFind, 0, inputNumbers.length - 1);
    }
}

module.exports = RecursiveBinarySearch;
